File: vbnet/ExampleAuthenticate.vb
Imports Tinkerforge

Module ExampleAuthenticate
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const SECRET As String = "My Authentication Secret!"

    Sub ConnectedCB(ByVal sender As IPConnection, _
                    ByVal connectReason As Short)
        Select Case connectReason
            Case IPConnection.CONNECT_REASON_REQUEST
                System.Console.WriteLine("Connected by request")
            Case IPConnection.CONNECT_REASON_AUTO_RECONNECT
                System.Console.WriteLine("Auto-Reconnect")
        End Select

        ' Authenticate first...
        Try
            sender.Authenticate(SECRET)
            System.Console.WriteLine("Authentication succeeded")
        catch
            System.Console.WriteLine("Could not authenticate")
            Exit Sub
        End Try

        ' ...reenable auto reconnect mechanism, as described below...
        sender.SetAutoReconnect(true)

        ' ...then trigger enumerate
        sender.Enumerate()
    End Sub

    Sub EnumerateCB(ByVal sender As IPConnection, _
                    ByVal uid As String, _
                    ByVal connectedUid As String, _
                    ByVal position As Char, _
                    ByVal hardwareVersion() As Short, _
                    ByVal firmwareVersion() As Short, _
                    ByVal deviceIdentifier As Integer, _
                    ByVal enumerationType As Short)
        System.Console.WriteLine("UID: {0}, Enumeration Type: {1}", uid, enumerationType)
    End Sub

    Sub Main()
        ' Create IPConnection and connect to brickd
        Dim ipcon As New IPConnection()

        ' Register Connected Callback
        AddHandler ipcon.Connected, AddressOf ConnectedCB

        ' Register Enumerate Callback
        AddHandler ipcon.EnumerateCallback, AddressOf EnumerateCB

        ' Disable auto reconnect mechanism, in case we have the wrong secret.
        ' If the authentication is successful, reenable it.
        ipcon.SetAutoReconnect(false)

        ' Connect to brickd
        ipcon.Connect(HOST, PORT)

        System.Console.WriteLine("Press key to exit")
        System.Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module

File: delphi/ExampleAuthenticate.pas
program ExampleAuthenticate;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, Device;

type
  TExample = class
  private
    ipcon: TIPConnection;
  public
    procedure ConnectedCB(sender: TIPConnection; const connectReason: byte);
    procedure EnumerateCB(sender: TIPConnection;
                          const uid: string; const connectedUid: string; const position: char;
                          const hardwareVersion: TVersionNumber; const firmwareVersion: TVersionNumber;
                          const deviceIdentifier: word; const enumerationType: byte);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  SECRET = 'My Authentication Secret!';

var
  e: TExample;

{ Authenticate each time the connection got (re-)established }
procedure TExample.ConnectedCB(sender: TIPConnection; const connectReason: byte);
begin
  case connectReason of
    IPCON_CONNECT_REASON_REQUEST:
    begin
      WriteLn('Connected by request');
    end;
    IPCON_CONNECT_REASON_AUTO_RECONNECT:
    begin
      WriteLn('Auto-Reconnect');
    end;
  end;
  { Authenticate first... }
  try
    sender.Authenticate(SECRET);
    WriteLn('Authentication succeeded');
  except
    WriteLn('Could not authenticate');
    exit;
  end;

  { ...reenable auto reconnect mechanism, as described below... }
  sender.SetAutoReconnect(true);

  { ...then trigger enumerate }
  sender.Enumerate;
end;

{ Print incoming enumeration }
procedure TExample.EnumerateCB(sender: TIPConnection;
                               const uid: string; const connectedUid: string; const position: char;
                               const hardwareVersion: TVersionNumber; const firmwareVersion: TVersionNumber;
                               const deviceIdentifier: word; const enumerationType: byte);
begin
  WriteLn('UID: ' + uid + ', Enumerate Type: ' + IntToStr(enumerationType));
end;

procedure TExample.Execute;
begin
  { Create IP Connection }
  ipcon := TIPConnection.Create;

  { Disable auto reconnect mechanism, in case we have the wrong secret. If the authentication is successful, reenable it. }
  ipcon.SetAutoReconnect(false);

  { Register connected callback to "ConnectedCB" }
  ipcon.OnConnected := {$ifdef FPC}@{$endif}ConnectedCB;

  { Register enumerate callback to "EnumerateCB" }
  ipcon.OnEnumerate := {$ifdef FPC}@{$endif}EnumerateCB;

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.
